ipa-core/src/helpers/transport/handler.rs

```diff
@@ -0,0 +1,201 @@
+use std::{
+    fmt::{Debug, Formatter},
+    future::Future,
+    marker::PhantomData,
+};
+
+use async_trait::async_trait;
+use serde::de::DeserializeOwned;
+use serde_json::json;
+
+use crate::{
+    error::BoxError,
+    helpers::{
+        query::PrepareQuery, transport::routing::Addr, BodyStream, HelperIdentity,
+        TransportIdentity,
+    },
+    query::{
+        NewQueryError, PrepareQueryError, ProtocolResult, QueryCompletionError, QueryInputError,
+        QueryStatus, QueryStatusError,
+    },
+};
+
+/// Represents some response sent from MPC helper acting on a given request. It is rudimental now
+/// because we sent everything as HTTP body, but it could evolve.
+///
+/// ## Performance
+/// This implementation is far from being optimal. Between HTTP and transport layer, there exists
+/// one round of serialization and deserialization to properly represent the types. It is not critical
+/// to address, because MPC helpers have to handle a constant number of requests per query. Note
+/// that all requests tagged with [`crate::helpers::transport::RouteId::Records`] are not routed
+/// through [`RequestHandler`], so there is no penalty.
```

Comment:
I feel like this situation might improve somewhat if we stopped using axum. At somewhat increased risk of returning an invalid response, we can just give the transport layer a `Vec<u8>` and a content type, and have it return those without considering whether the bytes are valid for the claimed content type.

Reply:
I would also like to have type safety. One of the challenges I faced before was lack of clear contract between API client and server. Here the situation is aggravated by having extra layers that don't support strong type safety. This is somewhat mitigated by e2e tests but it is easy to miss, if API is never hit.

Reply:
Definitely we should have type safety (in the sense of checking requests and responses against a well-defined schema) somewhere on the datapath. But also, somewhere along the path the data has to be in a raw form. What I was trying to suggest is that we think about where the transitions are and which communication layers are dealing with raw form vs. type-safe parsed form.

Reply:
let me create an issue for that